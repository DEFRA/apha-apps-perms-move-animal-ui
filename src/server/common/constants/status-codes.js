/**
 * @typedef {Record<string, number>} StatusCodes
 */
export const statusCodes = {
  ok: 200,
  noContent: 204,
  badRequest: 400,
  unauthorized: 401,
  forbidden: 403,
  notFound: 404,
  imATeapot: 418,
  redirect: 302,
  serverError: 500,
  badGateway: 502,
  gatewayTimeout: 504
}
