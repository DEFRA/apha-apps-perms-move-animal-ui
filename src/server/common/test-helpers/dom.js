import { JSDOM } from 'jsdom'

/**
 * @param {string} payload
 * @returns {Document}
 */
export const parseDocument = (payload) => new JSDOM(payload).window.document
