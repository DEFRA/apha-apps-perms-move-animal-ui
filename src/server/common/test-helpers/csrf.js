import { statusCodes } from '~/src/server/common/constants/status-codes.js'
/* global expect, test */

/**
 * @param {OutgoingHttpHeaders} headers
 * @param {string} name
 * @returns {string}
 */
const findCookie = (headers, name) => {
  const setCookie = headers['set-cookie']

  /** @type {string[] | undefined} */
  const cookies = typeof setCookie === 'string' ? [setCookie] : setCookie
  return cookies?.find((h) => h?.includes(`${name}=`)) ?? ''
}

/**
 * @param {string} cookieString
 */
const extractCookieValue = (cookieString) =>
  cookieString.split('=')[1].split(';')[0]

/**
 * @param {function(): Server} serverFn
 * @param {ServerInjectOptions} injectionOptions
 */
export const testCsrfProtectedPost = (serverFn, injectionOptions) =>
  test('should reject if experiencing an apparent csrf attack (they have their session cookie, but no hidden form value)', async () => {
    const { statusCode } = await serverFn().inject({
      ...injectionOptions,
      headers: {
        ...(injectionOptions.headers ?? {}),
        Cookie: 'crumb=csrf-value'
      }
    })

    expect(statusCode).toEqual(statusCodes.forbidden)
  })

/**
 * @param {function(): Server} serverFn
 * @param {ServerInjectOptions} injectionOptions
 */
export const testCsrfProtectedGet = (serverFn, injectionOptions) =>
  test('should contain a cookie with a set CSRF value, and contain a hidden form field to submit that CSRF back to the server', async () => {
    const { headers, payload } = await serverFn().inject(injectionOptions)

    const token = extractCookieValue(findCookie(headers, 'crumb'))
    expect(token).not.toHaveLength(0)
    expect(payload).toContain(
      `<input type="hidden" name="crumb" value="${token}"`
    )
  })

/**
 * @param {ServerInjectOptions} injectionOptions
 * @returns {ServerInjectOptions}
 */
export const withCsrfProtection = (injectionOptions) => ({
  ...injectionOptions,
  payload: {
    ...(typeof injectionOptions.payload === 'object'
      ? injectionOptions.payload
      : {}),
    crumb: 'csrf-value'
  },
  headers: {
    ...(injectionOptions.headers ?? {}),
    Cookie: 'crumb=csrf-value'
  }
})

/**
 * @import { OutgoingHttpHeaders } from 'http'
 * @import { ServerInjectOptions, Server } from '@hapi/hapi'
 */
