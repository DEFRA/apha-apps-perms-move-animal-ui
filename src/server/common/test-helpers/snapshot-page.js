import { parseDocument } from '../../common/test-helpers/dom.js'
import { createServer } from '~/src/server/index.js'
import { withCsrfProtection } from './csrf.js'
/** @import { Server } from '@hapi/hapi' */

/* global expect, it, beforeAll, afterAll, describe */

/**
 * @param {string} describeMessage
 * @param {string} shouldMessage
 * @param {string} urlPath
 */
export const describePageSnapshot = (
  describeMessage,
  shouldMessage,
  urlPath
) => {
  describe(describeMessage, () => {
    /** @type {Server} */
    let server

    beforeAll(async () => {
      server = await createServer()
      await server.initialize()
    })

    afterAll(async () => {
      await server.stop({ timeout: 0 })
    })

    it(shouldMessage, async () => {
      const { payload } = await server.inject(
        withCsrfProtection({
          method: 'GET',
          url: urlPath
        })
      )

      const content =
        parseDocument(payload).querySelector('#main-content')?.innerHTML

      expect(content).toMatchSnapshot()
    })
  })
}
