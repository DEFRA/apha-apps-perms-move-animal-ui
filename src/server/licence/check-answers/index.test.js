import { createServer } from '~/src/server/index.js'
import { statusCodes } from '~/src/server/common/constants/status-codes.js'
import { withCsrfProtection } from '~/src/server/common/test-helpers/csrf.js'
import { parseDocument } from '~/src/server/common/test-helpers/dom.js'
import SessionTestHelper from '../../common/test-helpers/session-helper.js'
import { licenceSummaryPage } from './index.js'
import { receiveMethodPage } from '../receiveMethod/index.js'
import { describePageSnapshot } from '../../common/test-helpers/snapshot-page.js'

const pageUrl = '/receiving-the-licence/check-answers'

const testName = 'test_name'
const testSurname = 'test_surname'
const emailMethod = 'email'
const postMethod = 'post'
const testEmail = 'test@email.com'

const defaultState = {
  fullName: {
    firstName: testName,
    lastName: testSurname
  },
  receiveMethod: emailMethod,
  emailAddress: testEmail
}

describe('#licenceSummaryPage', () => {
  /** @type {Server} */
  let server

  beforeAll(async () => {
    server = await createServer()
    await server.initialize()
  })

  afterAll(async () => {
    await server.stop({ timeout: 0 })
  })

  describe('email method selected', () => {
    /** @type {SessionTestHelper} */
    let session

    beforeEach(async () => {
      session = await SessionTestHelper.create(server)

      await session.setState('licence', defaultState)
    })

    it('should render expected response', async () => {
      const { payload, statusCode } = await server.inject(
        withCsrfProtection(
          {
            method: 'GET',
            url: pageUrl
          },
          {
            Cookie: session.sessionID
          }
        )
      )

      const document = parseDocument(payload)
      expect(document.title).toBe(licenceSummaryPage.pageTitle)
      expect(statusCode).toBe(statusCodes.ok)

      expect(payload).toEqual(
        expect.stringContaining(`${testName} ${testSurname}`)
      )
    })

    it('should redirect user to receive method page if they`ve selected Post', async () => {
      await session.setState('licence', {
        ...defaultState,
        receiveMethod: postMethod
      })

      const { statusCode, headers } = await server.inject(
        withCsrfProtection(
          {
            method: 'GET',
            url: pageUrl
          },
          {
            Cookie: session.sessionID
          }
        )
      )

      expect(statusCode).toBe(statusCodes.redirect)
      expect(statusCode).toBe(statusCodes.redirect)
      expect(headers.location).toBe(
        `${receiveMethodPage.urlPath}?redirect_uri=${licenceSummaryPage.urlPath}`
      )
    })
  })
})

describePageSnapshot({
  describes: '#licenceSummaryPage.content',
  it: 'should render the expected content',
  pageUrl
})

/**
 * @import { Server } from '@hapi/hapi'
 */
