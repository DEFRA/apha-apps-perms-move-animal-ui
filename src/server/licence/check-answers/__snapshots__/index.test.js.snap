// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#licenceSummaryPage.content should render the expected content 1`] = `
"
  <div class="govuk-grid-row">
    <div class="govuk-grid-column-two-thirds">
      <h1 class="govuk-heading-l">Check your answers before you continue your application</h1>
    </div>
  </div>
    <form method="POST" novalidate="">
      <div class="govuk-grid-row">
        <div class="govuk-grid-column-two-thirds">
          <input type="hidden" name="crumb" value="csrf-value">
          <input type="hidden" name="nextPage" value="">
  
  <dl class="govuk-summary-list govuk-!-margin-bottom-9">
  <div class="govuk-summary-list__row">
    <dt class="govuk-summary-list__key govuk-!-width-one-half govuk-!-font-weight-regular">
      What is the name of the County Parish Holding (CPH) owner?
    </dt>
    <dd class="govuk-summary-list__value">
      test_name test_surname
    </dd>
    <dd class="govuk-summary-list__actions">
      <a class="govuk-link" href="/receiving-the-licence/licence-name?redirect_uri=/receiving-the-licence/check-answers" data-testid="fullName-change-link">Change<span class="govuk-visually-hidden"> What is the name of the County Parish Holding (CPH) owner?</span></a>
    </dd>
  </div>
  <div class="govuk-summary-list__row">
    <dt class="govuk-summary-list__key govuk-!-width-one-half govuk-!-font-weight-regular">
      How would you like this licence sent to you?
    </dt>
    <dd class="govuk-summary-list__value">
      Email
    </dd>
    <dd class="govuk-summary-list__actions">
      <a class="govuk-link" href="/receiving-the-licence/licence-email-or-post?redirect_uri=/receiving-the-licence/check-answers" data-testid="receiveMethod-change-link">Change<span class="govuk-visually-hidden"> How would you like this licence sent to you?</span></a>
    </dd>
  </div>
  <div class="govuk-summary-list__row">
    <dt class="govuk-summary-list__key govuk-!-width-one-half govuk-!-font-weight-regular">
      What email address would you like the licence sent to?
    </dt>
    <dd class="govuk-summary-list__value">
      test@email.com
    </dd>
    <dd class="govuk-summary-list__actions">
      <a class="govuk-link" href="/receiving-the-licence/licence-enter-email-address?redirect_uri=/receiving-the-licence/check-answers" data-testid="emailAddress-change-link">Change<span class="govuk-visually-hidden"> What email address would you like the licence sent to?</span></a>
    </dd>
  </div>
</dl>

            <button type="submit" class="govuk-button" data-module="govuk-button" id="continue-button">
  Continue
</button>
        </div>
      </div>
    </form>
        "
`;
